shared/src/balancer_event_handler.rs

```diff
@@ -7,31 +7,132 @@ use crate::{
+    tokens: Option<Vec<H160>>,
+    specialization: Option<PoolSpecialization>,
+    block_created: u64,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub struct WeightedPool {
+    pool_id: H256,
+    pool_address: H160,
+    tokens: Vec<H160>,
+    specialization: PoolSpecialization,
+    block_created: u64,
+}
+
+impl TryInto<WeightedPool> for WeightedPoolBuilder {
+    type Error = ();
+
+    fn try_into(self) -> Result<WeightedPool, Self::Error> {
+        if self.pool_address.is_some() && self.tokens.is_some() {
+            return Ok(WeightedPool {
+                pool_id: self.pool_id,
+                pool_address: self.pool_address.unwrap(),
+                tokens: self.tokens.unwrap(),
+                specialization: self.specialization.unwrap(),
+                block_created: self.block_created,
+            });
+        }
+        Err(())
+    }
+}
+
+impl WeightedPoolBuilder {
+    fn update_from_event(&mut self, event: BalancerEvent) {
+        // Pool and Token Registration always occur in the same tx! So block_created doesn't need update.
+        // https://github.com/balancer-labs/balancer-v2-monorepo/blob/70843e6a61ad11208c1cfabf5cfe15be216ca8d3/pkg/pool-utils/contracts/BasePool.sol#L128-L130
+        match event {
+            BalancerEvent::PoolRegistered(pool_registration) => {
+                self.pool_address = Some(pool_registration.pool_address);
+                self.specialization = Some(pool_registration.specialization);
+            }
+            BalancerEvent::TokensRegistered(token_registration) => {
+                self.tokens = Some(token_registration.tokens)
+            }
+        }
+    }
+}
+
+impl From<(EventIndex, PoolRegistered)> for WeightedPoolBuilder {
+    fn from(event_data: (EventIndex, PoolRegistered)) -> Self {
+        Self {
+            pool_id: event_data.1.pool_id,
+            pool_address: Some(event_data.1.pool_address),
+            specialization: Some(event_data.1.specialization),
+            // Tokens not emitted with pool registration event.
+            tokens: None,
+            block_created: event_data.0.block_number,
+        }
+    }
+}
+
+impl From<(EventIndex, TokensRegistered)> for WeightedPoolBuilder {
+    fn from(event_data: (EventIndex, TokensRegistered)) -> Self {
+        Self {
+            pool_id: event_data.1.pool_id,
+            // pool_address and specialization not emitted with token registration event.
+            pool_address: None,
+            specialization: None,
+            tokens: Some(event_data.1.tokens),
+            block_created: event_data.0.block_number,
+        }
+    }
+}
+
+/// The BalancerPool struct represents in-memory storage of all deployed Balancer Pools
+#[derive(Debug, Default)]
+pub struct BalancerPools {
+    /// Used for O(1) access to all pool_ids for a given token
+    pools_by_token: HashMap<H160, HashSet<H256>>,
+    /// WeightedPool data for a given PoolId
+    pools: HashMap<H256, WeightedPoolBuilder>,
```

Comment:
I feel it would be cleaner to store in this hash map an `enum { WeightedPoolBuilder, WeightedPool }` or have two separate hash maps for each case.
When you get new events you build up the builder and when you know you have both events and can either change the enum or move the pool from one hash map into the other.
It might also be simpler (less code to write) to make WeightedPoolBuilder only store the two relevant events directly like
```rust
pub struct WeightedPoolBuilder {
    pool_registered: Option<PoolRegistered>,
    tokens_registered: Option<TokensRegistered>,
}
```
and then you know you can create the WeightedPool when both options are Some.

Reply:
I really like the second suggestion and will certainly go for it. As for keeping the two maps, I could see this also working but it will require some internal changes. Nice thing is that the pending pools (or so I imagine we will call it) will nearly always be empty.

Reply:
I admit I have always started working on this suggestion and it will really make a lot of thing much easier. It's going to be quite a big refactoring off what o have here so won't finish today, but it's super fun.

Reply:
I also really like the alternate `WeightedPoolBuilder` representation suggestion.

One thing I was thinking about is what happens if a re-org happens such that:
- Block 100: `PoolCreated` event
- Block 102: `TokensRegistered` event
- `replace_events(101)` - so we need to "undo half" of a weighted pool

I think for this, it would be easier to have a single hash map and keep track of event blocks separately instead of having a single `block_created`.

Reply:
Although remember that both events are emitted in the same transaction always.

Reply:
> Although remember that both events are emitted in the same transaction always. So we don't necessarily have to be that considerate.

Ah, cool, then its a non-issue.

In which case I do agree with @vkgnosis that two maps might be nicer.